apiserver/admin.go

```diff
@@ -118,83 +105,82 @@ func (a *admin) login(req params.LoginRequest, loginVersion int) (params.LoginRe
-		PublicDNSName: a.srv.publicDNSName(),
+	// apiRoot is the API root exposed to the client after login.
+	var apiRoot rpc.Root = newAPIRoot(
+		a.root.state,
+		a.srv.statePool,
+		a.srv.facades,
+		a.root.resources,
+		a.root,
+	)
+	apiRoot, err = restrictAPIRoot(
+		a.srv,
+		apiRoot,
+		a.root.model,
+		*authResult,
+	)
+	if err != nil {
+		return fail, errors.Trace(err)
 	}
 
-	var filters []facadeFilterFunc
+	var facadeFilters []facadeFilterFunc
+	var modelTag string
 	if authResult.anonymousLogin {
-		filters = append(filters, IsAnonymousFacade)
+		facadeFilters = append(facadeFilters, IsAnonymousFacade)
 	}
 	if authResult.controllerOnlyLogin {
-		loginResult.Facades = filterFacades(a.srv.facades, append(filters, IsControllerFacade)...)
-		apiRoot = restrictRoot(apiRoot, controllerFacadesOnly)
+		facadeFilters = append(facadeFilters, IsControllerFacade)
 	} else {
-		loginResult.ModelTag = model.Tag().String()
-		loginResult.Facades = filterFacades(a.srv.facades, append(filters, IsModelFacade)...)
-		apiRoot = restrictRoot(apiRoot, modelFacadesOnly)
+		facadeFilters = append(facadeFilters, IsModelFacade)
+		modelTag = a.root.model.Tag().String()
 	}
 
 	a.root.rpcConn.ServeRoot(apiRoot, serverError)
-
-	return loginResult, nil
+	return params.LoginResult{
+		Servers:       params.FromNetworkHostsPorts(hostPorts),
+		ControllerTag: a.root.model.ControllerTag().String(),
+		UserInfo:      authResult.userInfo,
+		ServerVersion: jujuversion.Current.String(),
+		PublicDNSName: a.srv.publicDNSName(),
+		ModelTag:      modelTag,
+		Facades:       filterFacades(a.srv.facades, facadeFilters...),
+	}, nil
 }
 
 type authResult struct {
-	anonymousLogin      bool
-	userLogin           bool
-	controllerOnlyLogin bool
-	userInfo            *params.AuthUserInfo
+	tag                    names.Tag // nil if external user login
+	anonymousLogin         bool
+	userLogin              bool // false if anonymous user
+	controllerOnlyLogin    bool
+	controllerMachineLogin bool
+	userInfo               *params.AuthUserInfo
 }
 
 func (a *admin) authenticate(req params.LoginRequest) (*authResult, error) {
 	result := &authResult{
 		controllerOnlyLogin: a.root.modelUUID == "",
 		userLogin:           true,
 	}
+	if req.AuthTag != "" {
+		tag, err := names.ParseTag(req.AuthTag)
+		if err != nil {
+			return nil, errors.Annotate(err, "could not parse auth tag")
+		}
+		result.tag = tag
+	}
 
 	// Maybe rate limit non-user auth attempts.
-	machineAgent := false
 	if req.AuthTag != "" {
-		var err error
-		kind, err := names.TagKind(req.AuthTag)
-		// Check for anonymous user login.
-		if kind == names.UserTagKind {
-			userTag, err := names.ParseUserTag(req.AuthTag)
-			if err != nil {
-				return nil, errors.Trace(err)
-			}
-			result.anonymousLogin = userTag.Id() == api.AnonymousUsername && len(req.Macaroons) == 0
-			result.userLogin = !result.anonymousLogin
+		tag, err := names.ParseTag(req.AuthTag)
+		if err == nil {
+			result.tag = tag
```

Comment:
Heh, yeah :)

I wrote the one above first, then realised we should rate-limit if the tag is bad -- per the old code -- to prevent faulty clients from hammering on the server. Fixed, thanks.